### PR TITLE
New option to preserve method labels always.

### PR DESCRIPTION
**Pull request by RackerWilliams**
A new configuration option preserveMethodLabels instructs api-checker
to always ensure that method labels are preserved in the state
machine.  If false, method labels may be removed to enable the
creation of smaller state machines.  The default is false.

### CODE REVIEW

**Comment by wdschei** [comment]
By making the default false, this will force a major version bump. Not a big deal, just a side effect of this change.

**Comment by wdschei** [comment]
Or is false mimicking current behavior and true fixes the corner cases?

**Comment by RackerWilliams** [comment]
Default false preserves the behavior that exists today.   Optimization stages drop the label if it means that they'll get a more optimal state machine.  They always have. ```false``` preserves that behavior.  With this new feature you can tell the opt stages that you want to keep the labels no matter what.  None of this affects validation of the api, BTW.  Just whether you're guaranteed to get a label. Currently if you run the opt stages you have no guarantee.

File: core/src/test/scala/com/rackspace/com/papi/components/checker/RaxRolesBehaviors.scala [hunk 1]
```diff
@@ -32,15 +32,40 @@ trait RaxRolesBehaviors {
   def configWithRolesEnabled =
     TestConfig(false, false, true, true, true, 1, true, true, true, "XalanC", true, true, true, true, false, false, true, true)
 
+  def configWithRolesEnabledAndPreservedLabels = { 
+    val tc = TestConfig(false, false, true, true, true, 1, true, true, true, "XalanC", true, true, true, true, false, false, true, true)
+    tc.preserveMethodLabels = true
+    tc
+  }
+
```
[the remark this answers is not in the document]
Comment (RackerWilliams) [reply core/src/test/scala/com/rackspace/com/papi/components/checker/RaxRolesBehaviors.scala]: Uh yea, I totally regret going that route.  Try to not do this on new tests, but it's awkward to modify old tests and not be consistent there.  Should go back and fix, but would like to write a script that will scan the scala file and fix it for me :-)

File: core/src/test/scala/com/rackspace/com/papi/components/checker/wadl/WADLCheckerRaxRolesSpec.scala [hunk 2]
```diff
@@ -147,6 +147,13 @@ class WADLCheckerRaxRolesSpec extends BaseCheckerSpec {
       tf.enableRaxRolesExtension = true
       tf
     }
+    val raxRolesEnabledRemoveDupsPreserveLabels  = {
+      val tf = TestConfig()
```
[the remark this answers is not in the document]
Comment (RackerWilliams) [reply core/src/test/scala/com/rackspace/com/papi/components/checker/wadl/WADLCheckerRaxRolesSpec.scala]: Gosh, no idea why I did that.

File: core/src/test/scala/com/rackspace/com/papi/components/checker/wadl/WADLCheckerRaxRolesSpec.scala [hunk 3]
```diff
@@ -161,6 +168,14 @@ class WADLCheckerRaxRolesSpec extends BaseCheckerSpec {
...
+      tf.removeDups = true
+      tf.enableRaxRolesExtension = true
+      tf.maskRaxRoles403 = true
+      tf
+    }
+
 
     val wadls = Map[String, NodeSeq]("A WADL with rax:roles but no references" -> raxRolesWADLNoRef,
                                      "A WADL with rax:roles and method references" -> raxRolesWADLRef)
```
[the remark this answers is not in the document]
Comment (RackerWilliams) [reply core/src/test/scala/com/rackspace/com/papi/components/checker/wadl/WADLCheckerRaxRolesSpec.scala]: I'll keep that in mind.

**Pull request merged**